PCF-ADR-0008-ray-cluster-authentication.md

```diff
@@ -0,0 +1,74 @@
+# Project CodeFlare - RayCluster Authentication
+|                |                                                                                  |
+| -------------- | -------------------------------------------------------------------------------- |
+| Date           | 12/19/2023                                                                       |
+| Scope          |                                                                                  |
+| Status         | provisional                                                                      |
+| Authors        | [Kevin Postlethwait](@KPostOffice)                                               |
+| Supersedes     | N/A                                                                              |
+| Superseded by: | N/A                                                                              |
+| Issues         |                                                                                  |
+| Other docs:    | none                                                                             |
+
+## What
+
+Enabling OpenShift OAuth as an identity provider and authorization mechanism for KubeRay RayClusters.
+
+## Why
+
+Currently there is no authentication by default enabled in front of the Ray Dashboard and we would like to offer a
+unified login experience for the whole of ODH by way of OpenShift OAuth
+
+## Goals
```

Comment:
There is the question of the RayJob support. The Kuberay RayJob controller communicates with the Ray cluster Job endpoint. Will that work, do we want to support that? If yes, could it be possible that the KubeRay serviced account be authorised to access that endpoint automatically?

Reply:
If it communicates using the kuberay service account then I think it should already have access since we will be checking for ability to Get RayClusters in a specific namespace

Reply:
Currently the client the operator uses to connect to the Ray cluster job endpoint is not authenticated. We have to work with the Kuberay team to enable authentication. In addition to that client, the job is actually submitted in a separate "submitter" Job, using the Ray CLI. So this has to be authenticated as well.

Then for the authorisation, as you suggested, using the Kuberay service account identity / token sounds like a logical approach.

Reply:
So I thought that Service would point to the oauth-proxy endpoint? I'd be inclined to think in multi-tenant environments, this would be a must-have that service be secured?

Reply:
I had been making a new Service object originally. KubeRay creates a service that maps to `client`, `dashboard`, `metrics`, `redis`, and `serve` ports on the cluster head. These can be changed by specifying the `headService` in the head node spec instead. Would each of the ports that we want to expose need their own proxies?

Reply:
My understanding is that the `dashboard` endpoint both serve the dashboard webapp and the job REST API. If that understanding is correct, the Kuberay operator won't be able to connect to that endpoint without authenticating, to check the status of the job via the REST API. This reasoning likely applies to the `client` endpoint, that we want to secure for interactive sessions, and the Kuberay operator uses to submit the job?

Reply:
That makes sense to me. I'll adjust the proposal to cover:

1. kuberay authenticating when submitting jobs
2. mutating webhook should also make sure that only services covered by OAuth proxies are created. (no default proxies)